### PR TITLE
BAH-2015 | Closing of open orders

### PR DESCRIPTION
**Pull request by Arjun-Go**
JIRA link - https://bahmni.atlassian.net/jira/software/c/projects/BAH/boards/35?modal=detail&selectedIssue=BAH-2015

Summary - To update fulfiller status as COMPLETED once a report is uploaded against a lab order

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java [hunk 1]
```diff
@@ -52,15 +52,19 @@ public class ObsBasedDiagnosticReportService extends BaseFhirService<DiagnosticR
 	
 	@Autowired
 	private DiagnosticReportObsValidator diagnosticReportObsValidator;
-
+	
 	@Autowired
 	private DiagnosticReportRequestValidator diagnosticReportRequestValidator;
 	
+	@Autowired
+	private OrderUpdateService orderUpdateService;
```
Comment (pajmera4) [inline api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java:60]: Where this orderUpdateService is used.

File: api/src/main/java/org/openmrs/module/fhirExtension/validators/DiagnosticReportRequestValidator.java [hunk 2]
```diff
@@ -28,7 +26,10 @@ private void validateBasedOn(FhirDiagnosticReport fhirDiagnosticReport) {
         Patient patient = fhirDiagnosticReport.getSubject();
         Integer conceptId = fhirDiagnosticReport.getCode().getId();
         List<Order> allOrders = orderService.getAllOrdersByPatient(patient);
-        long matchingOrdersCount = allOrders.stream().filter(Order::isActive).filter((order) -> order.getConcept().getId().equals(conceptId)).count();
+        long matchingOrdersCount = allOrders.stream()
+                .filter(order -> !Order.FulfillerStatus.COMPLETED.equals(order.getFulfillerStatus()))
+                .filter(order -> !order.getVoided())
+                .filter((order) -> order.getConcept().getId().equals(conceptId)).count();
```
Comment (pajmera4) [inline api/src/main/java/org/openmrs/module/fhirExtension/validators/DiagnosticReportRequestValidator.java:32]: So as per this implementation, it means that an order should always be present for a diagnostic report. 
As, If a report is uploaded without any order, we are verifying if the matching order is present or not.

File: api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java (hunk 1 above)
Comment (Arjun-Go) [reply api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java:58]: orderUpdateService is being used in create method of ObsBasedDiagnosticReportService class

File: api/src/main/java/org/openmrs/module/fhirExtension/validators/DiagnosticReportRequestValidator.java (hunk 2 above)
Comment (Arjun-Go) [reply api/src/main/java/org/openmrs/module/fhirExtension/validators/DiagnosticReportRequestValidator.java]: validation is done against the order

File: api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java [hunk 3]
```diff
@@ -52,22 +50,26 @@ public class ObsBasedDiagnosticReportService extends BaseFhirService<DiagnosticR
...
+	
 	@Override
 	public DiagnosticReport create(@Nonnull DiagnosticReport diagnosticReport) {
 		try {
 			FhirDiagnosticReport fhirDiagnosticReport = obsBasedDiagnosticReportTranslator.toOpenmrsType(diagnosticReport);
 			diagnosticReportObsValidator.validate(fhirDiagnosticReport);
-			if (diagnosticReport.getBasedOn().size() > 0 && diagnosticReport.getBasedOn().get(0).getReference() != null)
-				diagnosticReportRequestValidator.validate(fhirDiagnosticReport);
+			diagnosticReportRequestValidator.validate(diagnosticReport, fhirDiagnosticReport);
+			orderUpdateService.updateOrder(diagnosticReport, fhirDiagnosticReport);
```
Comment (pajmera4) [inline api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java:72]: Should we update the order after saving the diagnostic report. As in I am thinking of the scenario where after updating the order, there is some exception while creating/updating the diagnostic report.

File: api/src/main/java/org/openmrs/module/fhirExtension/service/OrderUpdateService.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            String orderUuid = diagnosticReport.getBasedOn().get(0).getIdentifier().getValue();
+            currentOrder = orderService.getOrderByUuid(orderUuid);
+        } else {
+            Patient patient = fhirDiagnosticReport.getSubject();
+            Integer conceptId = fhirDiagnosticReport.getCode().getId();
+            List<Order> allOrders = orderService.getAllOrdersByPatient(patient);
+            Optional<Order> optionalOrder = allOrders.stream()
+                    .filter(order -> !Order.FulfillerStatus.COMPLETED.equals(order.getFulfillerStatus()))
+                    .filter(order -> !order.getVoided())
+                    .filter((order) -> order.getConcept().getId().equals(conceptId))
```
Comment (pajmera4) [inline api/src/main/java/org/openmrs/module/fhirExtension/service/OrderUpdateService.java:36]: we can combine all filter predicates.

Comment (Arjun-Go) [reply api/src/main/java/org/openmrs/module/fhirExtension/service/OrderUpdateService.java:36]: From readability point of view, we felt that this is more appropriate than using predicates.

File: api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java (hunk 3 above)
Comment (Arjun-Go) [reply api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java]: Moved order update logic after creating/updating diagnostic report

File: api/src/test/java/org/openmrs/module/fhirExtension/service/OrderUpdateServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		
+		when(orderService.getOrderByUuid(orderUuid)).thenReturn(order);
+		
+		orderUpdateService.updateOrder(diagnosticReport, fhirDiagnosticReport);
+		
+		assertEquals(Order.FulfillerStatus.COMPLETED, order.getFulfillerStatus());
+	}
+	
+	@Test
+	public void shouldUpdateOrderFulfillerStatusWhenUploadingFromSidePanel() {
```
Comment (pajmera4) [inline api/src/test/java/org/openmrs/module/fhirExtension/service/OrderUpdateServiceTest.java:56]: Service does not know about side panel. So can we update the test case name here.

**Review by pajmera4 — APPROVED** [review]

File: api/src/test/java/org/openmrs/module/fhirExtension/service/OrderUpdateServiceTest.java (hunk 5 above)
Comment (Arjun-Go) [reply api/src/test/java/org/openmrs/module/fhirExtension/service/OrderUpdateServiceTest.java:56]: Sure, updating name to `shouldUpdateOrderFulfillerStatusWhenUploadingReportAgainstAPendingLabOrder`

**Pull request merged**